src/plugins/intel_gpu/src/plugin/transformations/fc_horizontal_fusion.cpp

```diff
@@ -0,0 +1,208 @@
+// Copyright (C) 2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "fc_horizontal_fusion.hpp"
+
+#include "intel_gpu/op/fully_connected.hpp"
+#include "intel_gpu/op/fully_connected_compressed.hpp"
+
+#include "openvino/core/rt_info.hpp"
+#include <openvino/opsets/opset1.hpp>
+#include "openvino/pass/pattern/op/or.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "transformations/utils/utils.hpp"
+#include "intel_gpu/op/placeholder.hpp"
+
+namespace ov {
+namespace intel_gpu {
+
+FullyConnectedHorizontalFusion::FullyConnectedHorizontalFusion() {
+    using namespace ov::pass::pattern;
+
+    auto is_target_pattern = [](const Output<Node>& output) {
+        // Currently this pass targets only compressed FCs (QKV) on dynamic generative models
+        // inputs: input, weight, bias, scale, [zp]
+        // Bias/scale/zp are constant or none
+        // if it is not constant, the only allowed cases are Constant => convert
+        // All FCs have same # of valid inputs (e.g., if one of the fc has zp, all fcs have zp)
+
+        auto is_constant = [](const std::shared_ptr<ov::Node> node) {
+            if (std::dynamic_pointer_cast<ov::op::v0::Constant>(node))
+                return true;
+            if (std::dynamic_pointer_cast<ov::op::v0::Convert>(node) && std::dynamic_pointer_cast<ov::op::v0::Constant>(node->get_input_node_shared_ptr(0)))
+                return true;
+            if (std::dynamic_pointer_cast<ov::op::v1::Transpose>(node) && std::dynamic_pointer_cast<ov::op::v0::Constant>(node->get_input_node_shared_ptr(0)))
+                return true;
+            return false;
+        };
+        auto is_placeholder = [](const std::shared_ptr<ov::Node> node) {
+            return std::dynamic_pointer_cast<op::Placeholder>(node);
+        };
+        // Three FCs connected to the same input
```

Comment:
Why did you put everything to FC predicate? Can't it be something like

```cpp
auto in = any_input();
auto fc1 = wrap_type<FC>(in, ...);
auto fc2 = wrap_type<FC>(in, ...);
auto fc3 = wrap_type<FC>(in, ...);
```
?

Reply:
At first I did that, but without root node, having those pattern was meaningless..  The pattern was hitting for single fc even though I added three fcs in the matcher pattern. So I just check one fc and pattern around it. Do you have any better idea for a sub network pattern which does not saturate to a root node?

Reply:
If the pattern above worked for single FC, then it could be a bug in pattern matcher. Do you have a test to reproduce the issue?

Reply:
Not now, will make afterwards.